### ISSUE TITLE
Some key combinations lock the game into doing one action

### ISSUE THREAD

**Issue by akien-mga** [OpenDungeons/OpenDungeons#68]
This is a follow-up of issue #49. Now that the chatbox has been fixed, I've noticed that some other key combinations trigger similar effects, i.e. the map rotates or translates unendingly. I do believe the issue was there before the chatbox was implemented though, it's probably simply another OIS related issue.

So, here are some key combinations that trigger unexpected effects, when each key is pressed quickly one after the other.
- _Q + AltGr_ sets the map rotating to the left, _E + AltGr_ is the counterpart.
- _A + AltGr_ sets the map translating to the left, _D + AltGr_ to the right, etc.
- For the rotations, _AltGr + <X>_ seems to be the counterpart to _<X> + AltGr_, e.g. _AltGr + Q_ rotates the map to the right. This is not the case for the translation keys though (WASD).

**Comment by akien-mga** [comment]
Note that this might have to do with the focus, since for example _Q + Alt+Tab_ also triggers the issue (i.e. the map starts spinning because of Q, then the focus is lost with Alt+Tab, so the game doesn't get the event saying that the user stopped holding Q).

**Comment by hwoarangmy** [comment]
I don't think this kind of issues where there before the chat. I've not really investigated but I think the problem comes from receiving a key in keyrelease and not in keypressed. The origin is that the keyevent given by OIS for keypressed and keyreleased is not the same (I'm talking about arg.text). As a hack, I've used a vector to save the keycodes and, in keyrelease, the keycode is removed from the list if it is there. If it is not, it is passed as a normal key. But I guess that in the bug you have seen we receive some weird event.

**Comment by hwoarangmy** [comment]
I'm not sure I reproduce the same bug. For example, when I press D then Altgr, it works (adds d to the chat message and no map translation). But when I do Altgr then D, nothing is added to the chat message and the map translates to the right. In fact, pressing Altgr + a key do not print the key. The only bug I could see is that when you press Altgr+A, nothing happens (we may expect map translation on the left). But for me it is really a minor bug.
If you have not the same behaviour, please tell me and we will see what we can do to add logs to see what is happening.

**Comment by akien-mga** [comment]
Note that the key combinations trigger the issue on the map, and not on the chat. You should press "E + AltGr" or "AltGr + E" while the chatbox is closed.

**Comment by akien-mga** [comment]
As I thought, this is not related to the chatbox. I built an old commit from before the big multiplayer + chatbox pull requests, and the issue is the same as described in comment 0.

**Comment by hwoarangmy** [comment]
Ok, I should stop reading at 1am, I had not seen it was not related to chat. Sorry for making you do useless tests

**Comment by akien-mga** [comment]
That was not a useless test, even if it wasn't happened **in** the chatbox, it might have been related to some changes you made to input handling when implementing the chat feature. Now we know that it isn't :-)

**Comment by Bertram25** [comment]
@hwoarangmy Hi,
I don't want to put any pressure on you for such a stupid Input library so feel free to tell. :)
Do you think this can be dealt with for the release, or do you know an easy workaround we can add for the release?
If not, and if you don't want to deal with this before the 0.4.9 release, we can always list that in the known bugs list at release time.

Regards,

**Comment by akien-mga** [comment]
Unless @hwoarangmy really wants to have a look at it before the release, I think we can set the milestone to later. As you can see in the issue description, the only way I found to reproduce the issue was to use some key combinations with AltGr, and it's not a key that use usually in video games.

Most users probably won't notice it unless they do some deep and evil beta testing as I did, so a mention in the errata should be enough for the 0.4.9 release :-)

**Comment by Bertram25** [comment]
Sounds fair. :) We have enough for the release anyway. ;)

**Comment by akien-mga** [comment]
FYI, this is still valid in the current development branch. No need to change the milestone though, it's a very minor issue IMO (who is going to use AltGr in-game anyway? :-p)

**Comment by Bertram25** [comment]
Btw, I still do think keys pushed along with alt gr should be released when their no alt-gr counter parts are released (as a workaround). I still don't know how we can make this look nice in the code, though.

**Comment by akien-mga** [comment]
Found another occurrence of this bug, so I'm not sure the two of them would have the same cause: When you toggle the chat box while pressing a key, the corresponding event will never be released. E.g., try to press the `Left` key and then `Return`, it will lock the game into moving the camera to the left.
